fix(type hints): improve type hints in ibis.common

binste opened the pull request:
As suggested by @kszucs in https://github.com/ibis-project/ibis/pull/6862#issuecomment-1676799118, I copied all type hint changes from #6852 into this PR. I also fixed a few more mypy errors but this PR does not yet fix all in `ibis.common`. I think I'm not familiar enough with Ibis to suggest changes for the remaining ones but happy to brainstorm and review!

Some but not all outstanding points:

- [ ] Subclasses of `Annotable` create `Attributes without a default cannot follow attributes with one` errors. See https://github.com/ibis-project/ibis/issues/6844#issuecomment-1676403220
- [ ] `MapSet` does not take any input arguments but methods such as `__and__` pass arguments to class (`self.__class__(intersection)`) which gives `error: Too many arguments for "MapSet"  [call-arg]` errors. Should `MapSet` be an abstract base class with abstract methods and/or define `__init__`?

kszucs commented:
Thanks for pulling out into a separate PR! 

I need to play with it locally before I can answer the questions. In the meantime, can you please rebase?

kszucs commented on ibis/common/graph.py:
```diff
@@ -13,8 +13,10 @@
 
-def _flatten_collections(node: Any, filter: type) -> Iterator[Node]:
+
+def _flatten_collections(node: Any, filter: type[N]) -> Iterator[N]:
```
Nice!

kszucs commented on ibis/common/patterns.py:
```diff
@@ -382,6 +386,8 @@ class Variable(Slotted, Builder):
     """
 
     __slots__ = ("name",)
+    # Used Any as don't know what type the name is
```
Is it most certainly restricted to strings, going to check that.

kszucs commented on ibis/common/patterns.py:
```diff
@@ -797,6 +823,8 @@ class InstanceOf(Slotted, Singleton, Pattern):
 
     __slots__ = ("type",)
+    # Used Any as not sure how to type this
+    type: Any
```
This is directly passed to isinsance as typeinfo. Probably doesn't worth the trouble since it can be: `type`, `tuple[type, ...]`, `UnionTypeType`

binste replied:
I now annotated it as `_ClassInfo` which I defined as:

```python
# Keep this alias in sync with unittest.case._ClassInfo
if sys.version_info >= (3, 10):
    _ClassInfo: TypeAlias = type | UnionType | tuple["_ClassInfo", ...]
else:
    _ClassInfo: TypeAlias = type | tuple[_ClassInfo, ...]
```

This is what is used in the [standard library typeshed](https://github.com/python/typeshed/blob/main/stdlib/builtins.pyi#L1467) for `isinstance`. What do you think?

kszucs commented on ibis/common/patterns.py:
```diff
@@ -840,6 +868,8 @@ class GenericInstanceOf(Slotted, Pattern):
 
     __slots__ = ("origin", "fields")
+    origin: type
+    fields: dict
```
Theoretically this should be `FrozenDict[str, Pattern]`

binste replied:
Done

kszucs commented on ibis/common/patterns.py:
```diff
@@ -1647,6 +1711,9 @@ class Topmost(Slotted, Pattern):
     __slots__ = ("pattern", "filter")
+    pattern: Pattern
+    # Used Any as not sure how to type this
+    filter: Any
```
`Any` is actually a pattern defined in this file, I assume these should be `AnyType`

binste replied:
Done

kszucs commented on ibis/common/typing.py:
```diff
@@ -125,8 +125,8 @@ def get_bound_typevars(obj: Any) -> dict[str, Any]:
     -------
-    dict[str, Any]
-        Mapping of type variable name to type.
+    dict[TypeVar, tuple[str, type]]
```
We should remove the type annotation from the docstring.

binste replied:
You mean remove all type annotations from the docstrings or specifically this one?

binste commented on ibis/common/patterns.py:
```diff
@@ -755,6 +779,8 @@ class TypeOf(Slotted, Pattern):
 
     __slots__ = ("type",)
+    # Used Any as not sure how to type this
+    type: Any
```
Seems like an obvious one in hindsight... thanks! ;)

binste commented on ibis/common/patterns.py:
```diff
@@ -974,6 +1007,9 @@ class GenericCoercedTo(Slotted, Pattern):
 
     __slots__ = ("origin", "params", "checker")
+    origin: type
+    params: FrozenDict
```
Looks good to me. Done

binste commented:
Rebased on master.

kszucs commented on ibis/common/annotations.py:
```diff
@@ -43,6 +44,9 @@ class Annotation:
     __slots__ = ("_pattern", "_default", "_typehint")
+    _pattern: Pattern | Callable | None
+    _default: Any
+    _typehint: Any
```
```suggestion
    _typehint: AnyType
```

kszucs commented on ibis/common/annotations.py:
```diff
@@ -43,6 +44,9 @@ class Annotation:
 
     __slots__ = ("_pattern", "_default", "_typehint")
+    _pattern: Pattern | Callable | None
+    _default: Any
```
```suggestion
    _default: AnyType
```

kszucs commented on ibis/common/graph.py:
```diff
@@ -158,7 +160,7 @@ def map(self, fn: Callable, filter: Optional[type] = None) -> dict[Node, Any]:
         """
         filter = filter or Node
-        results = {}
+        results: dict[Node, Any] = {}
```
```suggestion
        results: dict[Node, AnyType] = {}
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -24,17 +25,30 @@
+if TYPE_CHECKING:
+    from types import ModuleType
+
+if sys.version_info >= (3, 10):
```
Can you move these conditional imports (including _ClassInfo below) to `ibis.common.typing` so we place the typing-compatibility related code to a centralized place?

kszucs commented on ibis/common/patterns.py:
```diff
@@ -382,6 +396,8 @@ class Variable(Slotted, Builder):
 
     __slots__ = ("name",)
+    # Used Any as don't know what type the name is
+    name: Any
```
```suggestion
    name: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -527,6 +548,7 @@ class Is(Slotted, Pattern):
     """
 
     __slots__ = ("value",)
+    value: Any
```
```suggestion
    value: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -557,6 +579,8 @@ class Capture(Slotted, Pattern):
     """
 
     __slots__ = ("key", "pattern")
+    key: Any
```
```suggestion
    key: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -716,6 +747,7 @@ class EqualTo(Slotted, Pattern):
     """
 
     __slots__ = ("value",)
+    value: Any
```
```suggestion
    value: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -737,6 +769,8 @@ class Option(Slotted, Pattern):
 
     __slots__ = ("pattern", "default")
+    pattern: Pattern
+    default: Any
```
```suggestion
    default: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -755,6 +789,7 @@ class TypeOf(Slotted, Pattern):
     """Pattern that matches a value that is of a given type."""
 
     __slots__ = ("type",)
+    type: type
```
We will probably need to rename these properties due to:

```
ibis/common/patterns.py:1226: error: Variable "ibis.common.patterns.SequenceOf.type" is not valid as a type  [valid-type]
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -1111,6 +1160,8 @@ class Contains(Slotted, Pattern):
 
     __slots__ = ("needle",)
+    # Used Any as not sure how to type this
+    needle: Any
```
```suggestion
    needle: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -1381,6 +1445,9 @@ def match(self, value, context):
     __slots__ = ("args", "return_")
+    args: tuple
+    # Used Any as not sure how to type this
+    return_: Any
```
```suggestion
    return_: AnyType
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -1647,6 +1719,9 @@ class Topmost(Slotted, Pattern):
 
     __slots__ = ("pattern", "filter")
+    pattern: Pattern
+    # Used Any as not sure how to type this
```
```suggestion
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -1669,6 +1744,9 @@ class Innermost(Slotted, Pattern):
 
     __slots__ = ("pattern", "filter")
+    pattern: Pattern
+    # Used Any as not sure how to type this
```
```suggestion
```

kszucs commented on ibis/common/patterns.py:
```diff
@@ -1669,6 +1744,9 @@ class Innermost(Slotted, Pattern):
     __slots__ = ("pattern", "filter")
+    pattern: Pattern
+    # Used Any as not sure how to type this
+    filter: Any
```
```suggestion
    filter: AnyType
```

kszucs reviewed (commented):
Before: `Found 85 errors in 20 files (checked 27 source files)`
After: `Found 50 errors in 15 files (checked 27 source files)`

Solid improvement already! 

There are a couple of nits left, could you please squash the commits afterwards?

binste commented:
Thanks for the inputs and great to see the progress! Implemented all requested changes and squashed commits down to the first commit message.


What mypy command did you run that only gave you 50 errors? I get
```

Found 140 errors in 30 files (checked 27 source files)
```
with `mypy ibis/common --ignore-missing-imports` but I'm also missing a few stub packages.

kszucs reviewed (approved):
Great! Thanks @binste!

binste commented:
Hmm seems like we have a different setup. If I install sub packages, the errors explode to 987 errors. I followed the instructions for conda on Mac at https://ibis-project.org/community/contribute/01_environment/#support-matrix and then ran `mypy --install-types`.
